const problem with combination of view and reshape_view

The following code fails to compile:

```cpp
const xtensor<int, 1> a{1, 2, 3, 4};
auto aa = reshape_view(a, {4});
xtensor<int, 1> r =  view(aa, range(0,2));
```

with the error:
```cpp
xtensor/xstrided_view_base.hpp:558:25: error: invalid conversion from ‘const value_type* {aka const int*}’ to ‘std::enable_if_t<true, int*> {aka int*}’ [-fpermissive]
         return m_e.data();
```

It works if:

  - `a` is not `const`, or if
  - `aa` is declared explicitely as `const` (`const auto aa=`), or if
  - the `reshape_view` is not assigned to an explicit variable `view(reshape_view(a, {4}), range(0,2));`

xtl_version="0.6.4"
xtensor_version="0.20.5"
xsimd_version="7.2.1"

The issue might be related to https://github.com/QuantStack/xtensor/issues/1494

--- a/include/xtensor/xstrided_view_base.hpp
+++ b/include/xtensor/xstrided_view_base.hpp
@@ -162,10 +162,10 @@ namespace xt
         const storage_type& storage() const noexcept;
 
         template <class E = xexpression_type, class ST = storage_type>
-        std::enable_if_t<detail::provides_data_interface<E, ST>::value, value_type*>
+        std::enable_if_t<detail::provides_data_interface<E, ST>::value, pointer>
         data() noexcept;
         template <class E = xexpression_type, class ST = storage_type>
-        std::enable_if_t<detail::provides_data_interface<E, ST>::value, const value_type*>
+        std::enable_if_t<detail::provides_data_interface<E, ST>::value, const_pointer>
         data() const noexcept;
         size_type data_offset() const noexcept;
 
@@ -554,7 +554,7 @@ namespace xt
     template <class D>
     template <class E, class ST>
     inline auto xstrided_view_base<D>::data() noexcept ->
-        std::enable_if_t<detail::provides_data_interface<E, ST>::value, value_type*>
+        std::enable_if_t<detail::provides_data_interface<E, ST>::value, pointer>
     {
         return m_e.data();
     }
@@ -566,7 +566,7 @@ namespace xt
     template <class D>
     template <class E, class ST>
     inline auto xstrided_view_base<D>::data() const noexcept ->
-        std::enable_if_t<detail::provides_data_interface<E, ST>::value, const value_type*>
+        std::enable_if_t<detail::provides_data_interface<E, ST>::value, const_pointer>
     {
         return m_e.data();
     }

--- a/test/test_xstrided_view.cpp
+++ b/test/test_xstrided_view.cpp
@@ -787,4 +787,13 @@ namespace xt
         auto it3 = it2 + 0;
         EXPECT_EQ(*it2, *it3);
     }
+
+    TEST(xstrided_view, view_on_const)
+    {
+        const xtensor<int, 1> a = {1, 2, 3, 4};
+        auto v = strided_view(a, {all()});
+        auto d = v.data();
+        EXPECT_TRUE((std::is_same<decltype(d), const int*>::value));
+        EXPECT_EQ(d[0], 1);
+    }
 }

